fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2993,6 +2986,43 @@ public CompletableFuture<Boolean> markIndexReadable(@Nonnull String indexName) {
         return markIndexReadable(getRecordMetaData().getIndex(indexName));
     }
 
+    private CompletableFuture<Boolean> checkRangeAndUpdateIndexState(Index index, byte[] indexKey, Transaction tr, boolean allowUniquePending) {
+        // An extension function to reduce markIndexReadable's complexity
+        CompletableFuture<Optional<Range>> builtFuture = firstUnbuiltRange(index);
+        CompletableFuture<Optional<RecordIndexUniquenessViolation>> uniquenessFuture = whenAllIndexUniquenessCommitChecks(index)
+                .thenCompose(vignore -> scanUniquenessViolations(index, 1).first());
+        return CompletableFuture.allOf(builtFuture, uniquenessFuture).thenApply(vignore -> {
+            Optional<Range> firstUnbuilt = context.join(builtFuture);
+            Optional<RecordIndexUniquenessViolation> uniquenessViolation = context.join(uniquenessFuture);
+
+            if (firstUnbuilt.isPresent()) {
+                throw new IndexNotBuiltException("Attempted to make unbuilt index readable" , firstUnbuilt.get(),
+                        LogMessageKeys.INDEX_NAME, index.getName(),
+                        "unbuiltRangeBegin", ByteArrayUtil2.loggable(firstUnbuilt.get().begin),
+                        "unbuiltRangeEnd", ByteArrayUtil2.loggable(firstUnbuilt.get().end),
+                        subspaceProvider.logKey(), subspaceProvider.toString(context),
+                        LogMessageKeys.SUBSPACE_KEY, index.getSubspaceKey());
+            } else if (uniquenessViolation.isPresent()) {
+                if (allowUniquePending) {
+                    updateIndexState(index.getName(), indexKey, IndexState.READABLE_UNIQUE_PENDING);
+                    // keep the index build data until the uniqueness violations are resolved and the index is marked as readable
+                    return true;
```

Comment:
Good point. 
IIUC, currently the function returns false only if the index is already readable. Maybe, when setting the UP state the function should return an unconditional false (and update the javadoc to "return true if changed to READABLE, false otherwise"). 
I'm not sure how valuable the return code will be if it only indicates an arbitrary change.

Reply:
On a second thought, changed to match the javadoc - `will contain <code>true</code> if the store was modified`.